helix-core/src/main/java/org/apache/helix/manager/zk/CallbackEventExecutor.java

```diff
@@ -0,0 +1,122 @@
+package org.apache.helix.manager.zk;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+
+import org.apache.helix.HelixManager;
+import org.apache.helix.NotificationContext;
+import org.apache.helix.common.DedupEventBlockingQueue;
+import org.apache.helix.zookeeper.zkclient.exception.ZkInterruptedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Each batch mode is enabled, each CallbackHandler has a CallbackEventTPExecutor.
+ * It has a dedupe queue for pending call back event. Pending call back event will
+ * be submitted to a thread pool one at a time when
+ * 1. This is the first ever call back event for the callBackHandler, or
+ * 2. The previous call back event handling process is finished in thread pool.
+ */
+
+public class CallbackEventExecutor {
+  private static Logger logger = LoggerFactory.getLogger(CallbackHandler.class);
+
+  private DedupEventBlockingQueue<NotificationContext.Type, NotificationContext>
+      _callBackEventQueue;
+  private final HelixManager _manager;
+  private Future _futureCallBackProcessEvent = null;
+  private final ThreadPoolExecutor _threadPoolExecutor;
+
+  public CallbackEventExecutor(HelixManager manager) {
+    _callBackEventQueue = new DedupEventBlockingQueue<>();
+    _manager = manager;
+    _threadPoolExecutor = CallbackEventTPFactory.getOrCreateThreadPool(manager.hashCode());
+  }
+
+  class CallbackProcessor implements Runnable {
+    private CallbackHandler _handler;
+    protected String _processorName;
+    NotificationContext _event;
+
+    public CallbackProcessor(CallbackHandler handler, NotificationContext event) {
+      _processorName = _manager.getClusterName() + "CallbackProcessor@" + Integer
+          .toHexString(handler.hashCode());
+      _handler = handler;
+      _event = event;
+    }
+
+    @Override
+    public void run() {
+      try {
+        _handler.invoke(_event);
+      } catch (ZkInterruptedException e) {
+        logger.warn(_processorName + " thread caught a ZK connection interrupt", e);
+      } catch (ThreadDeath death) {
+        throw death;
```

Comment:
I think we need to "submitPendingHandleCallBackEventToManagerThreadPool" no matter what is the exception in the previous execution. So it should either be put in the finally block or this throw should be an error log too.